### PR TITLE
ESQL: Don't mutate the BoolQueryBuilder in plan

### PR DESCRIPTION
**Pull request by nik9000**
This modifies ESQL's `QueryBuilder` merging to stop it from mutating
`BoolQueryBuilder`s in place. It's more efficient when you can do that,
but only marginally so. Instead we create a shallow copy of the same
builder and mutate *that*. That lines up much better with the plan being
immutable objects. It should be!

The resulting queries that ESQL sends to lucene are the same here - we
just modify how we build them.

This should stop a fun class of bugs that can come up when we mutate the
query builders in multiple threads - because we *do* replan the query in
multiple threads. That's fine. So long as we shallow copy, like we do in
this PR.

### CODE REVIEW

**Comment by nik9000** [comment]
I thought about trying to move the mutation so I could be sure it's not done in multiple threads, but this seemed both easier and more paranoid. We really treat the plan as immutable. It should be. So we shouldn't mutate query builders inside it.

File: x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/Clusters.java [hunk 1]
```diff
@@ -8,15 +8,17 @@
 package org.elasticsearch.xpack.esql.qa.multi_node;
 
 import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.local.LocalClusterConfigProvider;
 import org.elasticsearch.test.cluster.local.distribution.DistributionType;
 
 public class Clusters {
-    public static ElasticsearchCluster testCluster() {
+    public static ElasticsearchCluster testCluster(LocalClusterConfigProvider configProvider) {
```
Comment (dnhatn) [inline x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/Clusters.java:15]: Is there a specific reason we need to make this change in this PR? I've tried to understand, but I think I am missing something.

**Review by dnhatn — APPROVED** [review]
LGTM. Thank you for the quick fix, Nik!

**Review by astefan — APPROVED** [review]
LGTM. Awesome

File: x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/Clusters.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/Clusters.java:15]: I was pulling other features "up" out of single node tests into the main tests. I thought I'd end up with us having a test for this already. We don't, but I figured the updates to the testing was ok to have along for the ride.

**Pull request merged**